File: src/main/menu.ts
import { app, Menu, shell, BrowserWindow, dialog } from 'electron'

export function createApplicationMenu(mainWindow: BrowserWindow): void {
  const isMac = process.platform === 'darwin'

  const template: Electron.MenuItemConstructorOptions[] = [
    // macOS App menu
    ...(isMac
      ? [
          {
            label: app.name,
            submenu: [
              {
                label: `About ${app.name}`,
                click: () => showAboutDialog(mainWindow)
              },
              { type: 'separator' as const },
              {
                label: 'Preferences...',
                accelerator: 'Cmd+,',
                click: () => {
                  mainWindow.webContents.send('navigate-to', '/settings')
                }
              },
              { type: 'separator' as const },
              { role: 'services' as const },
              { type: 'separator' as const },
              { role: 'hide' as const },
              { role: 'hideOthers' as const },
              { role: 'unhide' as const },
              { type: 'separator' as const },
              { role: 'quit' as const }
            ]
          }
        ]
      : []),

    // File menu
    {
      label: 'File',
      submenu: [
        {
          label: 'Export History...',
          accelerator: isMac ? 'Cmd+E' : 'Ctrl+E',
          click: () => {
            mainWindow.webContents.send('trigger-export')
          }
        },
        {
          label: 'Import History...',
          accelerator: isMac ? 'Cmd+I' : 'Ctrl+I',
          click: () => {
            mainWindow.webContents.send('trigger-import')
          }
        },
        { type: 'separator' as const },
        isMac
          ? { role: 'close' as const }
          : {
              label: 'Quit',
              accelerator: 'Ctrl+Q',
              click: () => {
                app.quit()
              }
            }
      ]
    },

    // Edit menu
    {
      label: 'Edit',
      submenu: [
        { role: 'undo' as const },
        { role: 'redo' as const },
        { type: 'separator' as const },
        { role: 'cut' as const },
        { role: 'copy' as const },
        { role: 'paste' as const },
        ...(isMac
          ? [
              { role: 'pasteAndMatchStyle' as const },
              { role: 'delete' as const },
              { role: 'selectAll' as const }
            ]
          : [{ role: 'delete' as const }, { type: 'separator' as const }, { role: 'selectAll' as const }])
      ]
    },

    // Tools menu
    {
      label: 'Tools',
      submenu: [
        {
          label: 'JSON Formatter',
          accelerator: isMac ? 'Cmd+1' : 'Ctrl+1',
          click: () => {
            mainWindow.webContents.send('navigate-to-tool', '/json')
          }
        },
        {
          label: 'JWT Decoder',
          accelerator: isMac ? 'Cmd+2' : 'Ctrl+2',
          click: () => {
            mainWindow.webContents.send('navigate-to-tool', '/jwt')
          }
        },
        {
          label: 'Base64 Converter',
          accelerator: isMac ? 'Cmd+3' : 'Ctrl+3',
          click: () => {
            mainWindow.webContents.send('navigate-to-tool', '/base64')
          }
        },
        {
          label: 'URL Encoder/Decoder',
          accelerator: isMac ? 'Cmd+4' : 'Ctrl+4',
          click: () => {
            mainWindow.webContents.send('navigate-to-tool', '/url')
          }
        },
        {
          label: 'Regex Tester',
          accelerator: isMac ? 'Cmd+5' : 'Ctrl+5',
          click: () => {
            mainWindow.webContents.send('navigate-to-tool', '/regex')
          }
        },
        {
          label: 'Text Diff',
          accelerator: isMac ? 'Cmd+6' : 'Ctrl+6',
          click: () => {
            mainWindow.webContents.send('navigate-to-tool', '/diff')
          }
        },
        {
          label: 'Hash Generator',
          accelerator: isMac ? 'Cmd+7' : 'Ctrl+7',
          click: () => {
            mainWindow.webContents.send('navigate-to-tool', '/hash')
          }
        },
        {
          label: 'UUID Generator',
          accelerator: isMac ? 'Cmd+8' : 'Ctrl+8',
          click: () => {
            mainWindow.webContents.send('navigate-to-tool', '/uuid')
          }
        },
        {
          label: 'Timestamp Converter',
          accelerator: isMac ? 'Cmd+9' : 'Ctrl+9',
          click: () => {
            mainWindow.webContents.send('navigate-to-tool', '/timestamp')
          }
        },
        { type: 'separator' as const },
        {
          label: 'All Tools...',
          accelerator: isMac ? 'Cmd+0' : 'Ctrl+0',
          click: () => {
            mainWindow.webContents.send('navigate-to-tool', '/')
          }
        }
      ]
    },

    // View menu
    {
      label: 'View',
      submenu: [
        {
          label: 'Toggle History Panel',
          accelerator: isMac ? 'Cmd+Shift+H' : 'Ctrl+Shift+H',
          click: () => {
            mainWindow.webContents.send('toggle-history-panel')
          }
        },
        { type: 'separator' as const },
        {
          label: 'Reload',
          accelerator: isMac ? 'Cmd+R' : 'Ctrl+R',
          click: () => {
            mainWindow.reload()
          }
        },
        {
          label: 'Toggle Developer Tools',
          accelerator: isMac ? 'Cmd+Option+I' : 'Ctrl+Shift+I',
          click: () => {
            mainWindow.webContents.toggleDevTools()
          }
        },
        { type: 'separator' as const },
        { role: 'resetZoom' as const },
        { role: 'zoomIn' as const },
        { role: 'zoomOut' as const },
        { type: 'separator' as const },
        { role: 'togglefullscreen' as const }
      ]
    },

    // Window menu
    {
      label: 'Window',
      submenu: [
        { role: 'minimize' as const },
        { role: 'zoom' as const },
        ...(isMac
          ? [
              { type: 'separator' as const },
              { role: 'front' as const },
              { type: 'separator' as const },
              { role: 'window' as const }
            ]
          : [{ role: 'close' as const }])
      ]
    },

    // Help menu
    {
      label: 'Help',
      submenu: [
        {
          label: 'Documentation',
          click: async () => {
            await shell.openExternal('https://github.com/yourusername/dev-utils-hub/wiki')
          }
        },
        {
          label: 'GitHub Repository',
          click: async () => {
            await shell.openExternal('https://github.com/yourusername/dev-utils-hub')
          }
        },
        {
          label: 'Report Issue',
          click: async () => {
            await shell.openExternal('https://github.com/yourusername/dev-utils-hub/issues/new')
          }
        },
        { type: 'separator' as const },
        {
          label: 'Check for Updates',
          enabled: false, // 향후 구현
          click: () => {
            dialog.showMessageBox(mainWindow, {
              type: 'info',
              title: 'Check for Updates',
              message: 'You are running the latest version!',
              detail: `Version ${app.getVersion()}`,
              buttons: ['OK']
            })
          }
        },
        ...(isMac
          ? []
          : [
              { type: 'separator' as const },
              {
                label: `About ${app.name}`,
                click: () => showAboutDialog(mainWindow)
              }
            ])
      ]
    }
  ]

  const menu = Menu.buildFromTemplate(template)
  Menu.setApplicationMenu(menu)
}

function showAboutDialog(window: BrowserWindow): void {
  dialog.showMessageBox(window, {
    type: 'info',
    title: `About ${app.name}`,
    message: app.name,
    detail: `Version: ${app.getVersion()}\nElectron: ${process.versions.electron}\nChrome: ${process.versions.chrome}\nNode.js: ${process.versions.node}\nV8: ${process.versions.v8}\n\nA collection of developer utility tools.`,
    buttons: ['OK']
  })
}

File: src-tauri/src/menu.rs
use tauri::menu::{MenuBuilder, MenuItemBuilder, PredefinedMenuItem, SubmenuBuilder};
use tauri::{AppHandle, Emitter, Manager, Wry};

pub fn create_menu(app: &AppHandle) -> tauri::Result<tauri::menu::Menu<Wry>> {
    // ── Edit menu items ───────────────────────────────────────────────────────
    let undo = PredefinedMenuItem::undo(app, None)?;
    let redo = PredefinedMenuItem::redo(app, None)?;
    let cut = PredefinedMenuItem::cut(app, None)?;
    let copy = PredefinedMenuItem::copy(app, None)?;
    let paste = PredefinedMenuItem::paste(app, None)?;
    let select_all = PredefinedMenuItem::select_all(app, None)?;
    let separator = PredefinedMenuItem::separator(app)?;

    // ── File menu ─────────────────────────────────────────────────────────────
    let export_history = MenuItemBuilder::new("Export History...")
        .accelerator("CmdOrCtrl+E")
        .id("export-history")
        .build(app)?;

    let import_history = MenuItemBuilder::new("Import History...")
        .accelerator("CmdOrCtrl+I")
        .id("import-history")
        .build(app)?;

    #[cfg(not(target_os = "macos"))]
    let file_sep = PredefinedMenuItem::separator(app)?;

    #[cfg(not(target_os = "macos"))]
    let file_quit = MenuItemBuilder::new("Quit")
        .accelerator("CmdOrCtrl+Q")
        .id("file-quit")
        .build(app)?;

    #[cfg(not(target_os = "macos"))]
    let file_menu = SubmenuBuilder::new(app, "File")
        .item(&export_history)
        .item(&import_history)
        .item(&file_sep)
        .item(&file_quit)
        .build()?;

    #[cfg(target_os = "macos")]
    let file_menu = SubmenuBuilder::new(app, "File")
        .item(&export_history)
        .item(&import_history)
        .build()?;

    // ── Edit menu ─────────────────────────────────────────────────────────────
    let edit_menu = SubmenuBuilder::new(app, "Edit")
        .item(&undo)
        .item(&redo)
        .item(&separator)
        .item(&cut)
        .item(&copy)
        .item(&paste)
        .item(&select_all)
        .build()?;

    // ── View menu ─────────────────────────────────────────────────────────────
    let toggle_history = MenuItemBuilder::new("Toggle History")
        .accelerator("CmdOrCtrl+Shift+H")
        .id("toggle-history")
        .build(app)?;

    let toggle_sidebar = MenuItemBuilder::new("Toggle Sidebar")
        .accelerator("CmdOrCtrl+B")
        .id("toggle-sidebar")
        .build(app)?;

    let view_sep = PredefinedMenuItem::separator(app)?;
    let reload = PredefinedMenuItem::close_window(app, Some("Reload"))?;

    let devtools = MenuItemBuilder::new("Toggle DevTools")
        .accelerator("CmdOrCtrl+Shift+I")
        .id("devtools")
        .build(app)?;

    let view_menu = SubmenuBuilder::new(app, "View")
        .item(&toggle_history)
        .item(&toggle_sidebar)
        .item(&view_sep)
        .item(&reload)
        .item(&devtools)
        .build()?;

    // ── Help menu ─────────────────────────────────────────────────────────────
    let github = MenuItemBuilder::new("GitHub")
        .id("help-github")
        .build(app)?;

    let report_issue = MenuItemBuilder::new("Report Issue")
        .id("help-report-issue")
        .build(app)?;

    #[cfg(not(target_os = "macos"))]
    let help_sep = PredefinedMenuItem::separator(app)?;

    #[cfg(not(target_os = "macos"))]
    let help_about = MenuItemBuilder::new("About Dev Utils Hub")
        .id("help-about")
        .build(app)?;

    #[cfg(not(target_os = "macos"))]
    let help_menu = SubmenuBuilder::new(app, "Help")
        .item(&github)
        .item(&report_issue)
        .item(&help_sep)
        .item(&help_about)
        .build()?;

    #[cfg(target_os = "macos")]
    let help_menu = SubmenuBuilder::new(app, "Help")
        .item(&github)
        .item(&report_issue)
        .build()?;

    // ── App menu (macOS only) ─────────────────────────────────────────────────
    #[cfg(target_os = "macos")]
    let preferences = MenuItemBuilder::new("Preferences...")
        .accelerator("Cmd+,")
        .id("preferences")
        .build(app)?;

    #[cfg(target_os = "macos")]
    let app_sep1 = PredefinedMenuItem::separator(app)?;
    #[cfg(target_os = "macos")]
    let services = PredefinedMenuItem::services(app, None)?;
    #[cfg(target_os = "macos")]
    let app_sep2 = PredefinedMenuItem::separator(app)?;
    #[cfg(target_os = "macos")]
    let hide = PredefinedMenuItem::hide(app, None)?;
    #[cfg(target_os = "macos")]
    let app_sep3 = PredefinedMenuItem::separator(app)?;
    #[cfg(target_os = "macos")]
    let quit = PredefinedMenuItem::quit(app, None)?;

    #[cfg(target_os = "macos")]
    let app_menu = SubmenuBuilder::new(app, "Dev Utils Hub")
        .about(None)
        .item(&app_sep1)
        .item(&preferences)
        .item(&app_sep2)
        .item(&services)
        .item(&app_sep3)
        .item(&hide)
        .item(&quit)
        .build()?;

    // ── Assemble menu ─────────────────────────────────────────────────────────
    #[cfg(target_os = "macos")]
    let menu = MenuBuilder::new(app)
        .item(&app_menu)
        .item(&file_menu)
        .item(&edit_menu)
        .item(&view_menu)
        .item(&help_menu)
        .build()?;

    #[cfg(not(target_os = "macos"))]
    let menu = MenuBuilder::new(app)
        .item(&file_menu)
        .item(&edit_menu)
        .item(&view_menu)
        .item(&help_menu)
        .build()?;

    Ok(menu)
}

pub fn handle_menu_event(app: &AppHandle, event: &tauri::menu::MenuEvent) {
    match event.id().as_ref() {
        "preferences" => {
            app.emit("navigate-to", "/settings").ok();
        }
        "export-history" => {
            app.emit("trigger-export", ()).ok();
        }
        "import-history" => {
            app.emit("trigger-import", ()).ok();
        }
        "toggle-history" => {
            app.emit("toggle-history-panel", ()).ok();
        }
        "toggle-sidebar" => {
            app.emit("shortcut:toggle-history", ()).ok();
        }
        "devtools" => {
            if let Some(window) = app.get_webview_window("main") {
                window.open_devtools();
            }
        }
        "help-github" => {
            let _ = tauri_plugin_opener::open_url(
                "https://github.com/jellpd/dev-utils-hub",
                None::<&str>,
            );
        }
        "help-report-issue" => {
            let _ = tauri_plugin_opener::open_url(
                "https://github.com/jellpd/dev-utils-hub/issues",
                None::<&str>,
            );
        }
        #[cfg(not(target_os = "macos"))]
        "file-quit" => {
            app.exit(0);
        }
        _ => {}
    }
}
